File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/MethodRouter.java
package cn.wzbos.android.rudolph.router;

import android.app.Application;
import android.content.Context;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import cn.wzbos.android.rudolph.Consts;
import cn.wzbos.android.rudolph.RLog;
import cn.wzbos.android.rudolph.Rudolph;
import cn.wzbos.android.rudolph.annotations.Route;
import cn.wzbos.android.rudolph.exception.ErrorCode;
import cn.wzbos.android.rudolph.exception.ErrorMessage;
import cn.wzbos.android.rudolph.exception.RudolphException;
import cn.wzbos.android.rudolph.utils.TypeUtils;

public class MethodRouter extends Router<Object> {

    private Map<String, String> uriKvs;

    MethodRouter(UriRouter.Builder<?> builder) {
        super(builder);
        this.uriKvs = builder.getUriAllParams();
    }

    public Object open() {
        return open(null);
    }

    public Object open(Context context) {
        if (super.intercept(context))
            return null;

        if (target == null) {
            if (callback != null)
                callback.onError(this, new RudolphException(ErrorCode.NOT_FOUND, ErrorMessage.NOT_FOUND_ERROR));
            return null;
        }

        try {
            Method[] methods = this.target.getMethods();

            for (Method method : methods) {

                Route route = method.getAnnotation(Route.class);
                if (route != null && route.value().equalsIgnoreCase(this.routePath)) {
                    List<Object> values = new ArrayList<>();

                    Map<String, String> params = this.queryParameters;

                    if (null != params && params.size() > 0) {

                        for (Map.Entry<String, String> kv : params.entrySet()) {
                            String argName = kv.getKey();
                            String argType = kv.getValue();
                            Object argValue = null;
                            if (Consts.RAW_URI.equals(argName)) {
                                argValue = rawUrl;
                            } else {
                                if (Application.class.getName().equals(kv.getValue())) {
                                    argValue = Rudolph.getApplication();
                                } else if (Context.class.getName().equals(kv.getValue())) {
                                    argValue = (null != context) ? context : Rudolph.getApplication();
                                } else {
                                    for (Map.Entry<String, String> uriKv : uriKvs.entrySet()) {
                                        if (argName.equalsIgnoreCase(uriKv.getKey())) {
                                            argValue = TypeUtils.getObject(context, argName, uriKv.getValue(), argType);
                                            break;
                                        }
                                    }
                                }
                            }

                            values.add(argValue);
                        }
                    }

                    Object obj = method.invoke(null, values.toArray());

                    if (null != callback)
                        callback.onSuccess(this);

                    return obj;
                }
            }
        } catch (Exception e) {
            if (null != callback) {
                callback.onError(this, new RudolphException(ErrorCode.METHOD_INVOKE_FAILED, "方法调用异常!", e));
            } else {
                RLog.e("MethodRouter", "方法调用异常!", e);
            }
        }
        return null;
    }

}

File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/MethodRouter.kt
package cn.wzbos.android.rudolph.router

import android.content.Context
import androidx.fragment.app.Fragment
import cn.wzbos.android.rudolph.UnknownExtraType
import cn.wzbos.android.rudolph.annotations.Route
import cn.wzbos.android.rudolph.exception.ErrorCode
import cn.wzbos.android.rudolph.exception.ErrorMessage
import cn.wzbos.android.rudolph.exception.RudolphException
import cn.wzbos.android.rudolph.logger.RLog
import cn.wzbos.android.rudolph.utils.TypeUtils
import cn.wzbos.android.rudolph.utils.match

class MethodRouter internal constructor(builder: UriRouter.Builder<*>) : Router<Any?>(builder) {
    companion object {
        const val TAG = "MethodRouter"
    }

    private val uriKvs: MutableMap<String, String?> = builder.uriAllParams

    @Deprecated(message = "use execute", replaceWith = ReplaceWith("execute(context)"))
    fun open(context: Context?): Any? {
        return invoke(context = context)
    }

    override fun execute(): Any? {
        return invoke()
    }

    fun execute(fragment: Fragment?): Any? {
        return invoke(fragment = fragment)
    }

    fun execute(context: Context?): Any? {
        return invoke(context = context)
    }

    private fun invoke(
        context: Context? = null,
        fragment: Fragment? = null
    ): Any? {
        if (super.intercept(fragment ?: context)) return null
        if (target == null) {
            callback?.onError(
                this,
                RudolphException(ErrorCode.NOT_FOUND, ErrorMessage.NOT_FOUND_ERROR)
            )
            return null
        }
        try {
            target?.methods?.forEach { m ->
                val route = m.getAnnotation(Route::class.java)
                if (route != null && route.match(rawUrl)) {
                    val values: MutableList<Any?> = ArrayList()
                    extraTypes?.onEachIndexed { index, entry ->
                        //兼容2.1.0之前的版本，反射获取参数类型
                        if (entry.value is UnknownExtraType) {
                            entry.value.type = m.parameterTypes[index]
                        }
                    }?.forEach { (argName, argType) ->
                        var value: String? = null
                        for (element in uriKvs) {
                            if (argName.equals(element.key, ignoreCase = true)) {
                                value = element.value
                                break
                            }
                        }

                        val argValue: Any? = TypeUtils.getObject(
                            context,
                            fragment,
                            argName,
                            value,
                            argType
                        )
                        values.add(argValue)
                    }
                    val obj = m.invoke(null, *values.toTypedArray())
                    callback?.onSuccess(this)
                    return obj
                }
            }
        } catch (e: Exception) {
            if (null != callback) {
                callback?.onError(
                    this,
                    RudolphException(ErrorCode.METHOD_INVOKE_FAILED, "方法调用异常!", e)
                )
            } else {
                RLog.e(TAG, "方法调用异常!", e)
            }
        }
        return null
    }

}